src/anomalib/models/efficientad/torch_model.py

```diff
@@ -0,0 +1,350 @@
+        self.ae: AutoEncoder = AutoEncoder(out_channels=teacher_out_channels, padding=padding)
+        self.teacher_out_channels: int = teacher_out_channels
+
+        self.mean_std: nn.ParameterDict = nn.ParameterDict(
+            {
+                "mean": torch.zeros((1, self.teacher_out_channels, 1, 1)),
+                "std": torch.zeros((1, self.teacher_out_channels, 1, 1)),
+            }
+        )
+
+        self.quantiles: nn.ParameterDict = nn.ParameterDict(
+            {
+                "qa_st": torch.tensor(0.0),
+                "qb_st": torch.tensor(0.0),
+                "qa_ae": torch.tensor(0.0),
+                "qb_ae": torch.tensor(0.0),
+            }
+        )
+
+    def is_set(self, p_dic: nn.ParameterDict) -> bool:
+        for _, value in p_dic.items():
+            if value.sum() != 0:
+                return True
+        return False
+
+    def choose_random_aug_image(self, image: Tensor) -> Tensor:
+        aug_index = random.choice([1, 2, 3])
+        # Sample an augmentation coefficient λ from the uniform distribution U(0.8, 1.2)
+        coefficient = random.uniform(0.8, 1.2)
+        if aug_index == 1:
+            img_aug = transforms.functional.adjust_brightness(image, coefficient)
+        elif aug_index == 2:
+            img_aug = transforms.functional.adjust_contrast(image, coefficient)
+        elif aug_index == 3:
+            img_aug = transforms.functional.adjust_saturation(image, coefficient)
+        return img_aug
+
+    def forward(self, batch: Tensor, batch_imagenet: Tensor = None) -> Union[Tensor, dict]:
+        """Prediction by EfficientAD models.
+
+        Args:
+            batch (Tensor): Input images.
+
+        Returns:
+            Tensor: Predictions
+        """
+        with torch.no_grad():
+            teacher_output = self.teacher(batch)
+            if self.is_set(self.mean_std):
+                teacher_output = (teacher_output - self.mean_std["mean"]) / self.mean_std["std"]
+            _, c, h, w = teacher_output.shape
+
+        student_output = self.student(batch)
+        ae_output = self.ae(batch)
+        # 3: Split the student output into Y ST ∈ R 384×64×64 and Y STAE ∈ R 384×64×64 as above
+        student_output = student_output[:, : self.teacher_out_channels, :, :]
+        student_output_ae = student_output[:, -self.teacher_out_channels :, :, :]
+
+        distance_st = torch.pow(teacher_output - student_output, 2)
+
+        if self.training:
+            # Student loss
+            d_hard = torch.quantile(distance_st, 0.999)
+            loss_hard = torch.mean(distance_st[distance_st >= d_hard])
+            student_imagenet_output = self.student(batch_imagenet)[:, :self.teacher_out_channels, :, :]
+            loss_st = loss_hard + (1 / (c * h * w)) * torch.sum(torch.pow(student_imagenet_output, 2))
+
+            # Autoencoder and Student AE Loss
+            aug_img = self.choose_random_aug_image(batch)
```

Comment:
If the have imagenet normalization, then we must undo imagenet normalization, augment with PIL and redo imagenet normalization. Like this:


```python
class EfficientAdModel(nn.Module):
    def __init__(
        self,
        model_size: Literal['small', 'medium']):
        super().__init__()
        
        # create models
        self._out_channels = 384
        if size == 'small':
            self.teacher = get_pdn_small(self._out_channels)
            self.student = get_pdn_small(2 * self._out_channels)
        elif size == 'medium':
            self.teacher = get_pdn_medium(self._out_channels)
            self.student = get_pdn_medium(2 * self._out_channels)
        else:
            raise Exception()
        # state_dict = torch.load(config.weights, map_location='cpu')
        # teacher.load_state_dict(state_dict)
        self.teacher_norm = nn.BatchNorm2d(num_features=self._out_channels, momentum=None, affine=False)
        self.autoencoder = get_autoencoder(self._out_channels)
        
        self.autoencoder_transform = transforms.Compose([
            # undo the imagenet normalization
            transforms.Normalize(mean = [0., 0., 0.],
                                 std = [1/0.229, 1/0.224, 1/0.225]),
            transforms.Normalize(mean = [-0.485, -0.456, -0.406],
                                 std = [ 1., 1., 1. ]),
            transforms.ToPILImage(),
            # augmentation
            transforms.RandomChoice([
                transforms.ColorJitter(brightness=0.2),
                transforms.ColorJitter(contrast=0.2),
                transforms.ColorJitter(saturation=0.2)
            ]),
            # back to pytorch
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])        
        ])
        
    def forward(self, images: Tensor) -> tuple[Tensor, Tensor]:
        with torch.no_grad():
            teacher_output = self.teacher(images)
            # update the teacher normalization for the first 100 batches.
            if self.training and self.teacher_norm.num_batches_tracked < 100:
                self.teacher_norm.train()
            else:
                self.teacher_norm.eval()
            teacher_output = self.teacher_norm(teacher_output)
            
        student_output = self.student(images)
        
        if self.training:
            # during training, augment the image for the autoencoder and apply
            # the teacher and student to the augmented image also.
            images_for_ae = self.autoencoder_transform(images)
            with torch.no_grad():
                teacher_output_for_ae = self.teacher(images_for_ae)
                # do not update teacher normalization on augmented
                self.teacher_norm.eval()
                teacher_output_for_ae = self.teacher_norm(teacher_output_for_ae)
            student_output_for_ae = self.student(images_for_ae)
        else:
            # during inference, no augmentation so directly use teacher and student output
            images_for_ae = images
            teacher_output_for_ae = teacher_output
            student_output_for_ae = student_output
        autoencoder_output = self.autoencoder(images_for_ae)
        
        # split the student output
        student_output = student_output[:, :self._out_channels]
        student_output_for_ae = student_output_for_ae[:, self._out_channels:]
        
        # compute losses for training, anomaly maps for inference
        distance_st = (teacher_output - student_output) ** 2
        distance_stae = (autoencoder_output - student_output_for_ae)**2
        if self.training:
            # student loss
            d_hard = torch.quantile(distance_st, q=0.999)
            loss_hard = torch.mean(distance_st[distance_st >= d_hard])
            
            # autoencoder loss
            distance_ae = (teacher_output_for_ae - autoencoder_output) ** 2
            loss_ae = torch.mean(distance_ae)
            
            # autoencoder student loss
            loss_stae = torch.mean(distance_stae)
            
            # return losses
            output = {'loss_student_hard': loss_hard,
                      'loss_autoencoder': loss_ae,
                      'loss_autoencoder_student': loss_stae}
        else:
            # compute anomaly map of student and anomaly map of autoencoder
            maps_st = torch.mean(distance_st, dim=1, keepdim=True)
            maps_ae = torch.mean(distance_stae, dim=1, keepdim=True)
            
            # merge maps
            anomaly_maps = 0.5 * maps_st + 0.5 * maps_ae
            anomaly_maps = torch.nn.functional.pad(anomaly_maps, (4, 4, 4, 4))
            anomaly_scores = torch.amax(anomaly_maps, dim=(1, 2, 3))
            output = anomaly_maps, anomaly_scores
        return output
```

Reply:
Oh right great point! I'm trying to find another fix for this problem though

Reply:
I moved the imagenet normalization to the model definition